tests/ui/lint/invalid_from_utf8.rs

```diff
@@ -128,18 +128,21 @@ pub fn from_utf8() {
 }
 
 pub fn from_utf8_with_indirections() {
-    let mut a = [99, 108, 130, 105, 112, 112, 121];
-    std::str::from_utf8_mut(&mut a);
-    //~^ WARN calls to `std::str::from_utf8_mut`
-    str::from_utf8_mut(&mut a);
-    //~^ WARN calls to `str::from_utf8_mut`
-    let mut b = &mut a;
-    let mut c = b;
-    std::str::from_utf8_mut(c);
-    //~^ WARN calls to `std::str::from_utf8_mut`
-    str::from_utf8_mut(c);
-    //~^ WARN calls to `str::from_utf8_mut`
-    let mut c = &[99, 108, 130, 105, 112, 112, 121];
+    // NOTE: We used to lint on the patterns below, but due to the
+    // binding being mutable it could be changed between the
+    // decleration and the call and that would have created a
+    // false-positive, so until we can reliably avoid those false
+    // postive we don't lint on them. Example of FP below.
```

Comment:
Obviously, this is out of scope for this PR, but I'm curious why this PR is implemented as a HIR analysis instead of a MIR analysis? We have a dataflow framework for MIR that can answer these kinds of questions ("do we definitely know the value of this variable at this point in the program?") and it seems like it would be much simpler to write this as a MIR pass.

Reply:
The `expr_or_init` method came from clippy, same for `invalid_from_utf8` which also came from clippy and both of them were using the HIR, so there is a bit a historical precedent.

As for implementing the lint as a MIR lint, it could probably be done, but I also remember MIR lints being more tricky to do, in part because MIR is much less like surface Rust than MIR is but also because the cargo check/build difference of MIR lints makes MIR lints less attractive; I don't know if it would apply here. I will have to check it.